File: implementations/csharp/tests/Skight.AgentPlatform.Tests/Phase2BusinessWorkflowSpecs.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;
using Xunit;
using FluentAssertions;
using Skight.AgentPlatform;

namespace Skight.AgentPlatform.Tests
{
    /// <summary>
    /// BDD Specification Suite: Core Business Workflows (Phase 2)
    /// 
    /// These tests serve as executable documentation for the business capabilities of the Agent Platform.
    /// A domain expert or reviewer should be able to read these scenarios to understand all platform workflows 
    /// without inspecting internal implementation details.
    /// </summary>
    public class Phase2BusinessWorkflowSpecs
    {
        [Fact]
        public async Task Feature_GitAutomation_Stages_Commits_And_Pushes_Workspace_Changes()
        {
            // SCENARIO: Developer requests automated Git status, commit, and push operations.
            // GIVEN: A workspace directory initialized as a Git repository with pending changes.
            var testRepoDir = Path.Combine(Path.GetTempPath(), "test_git_repo_" + Guid.NewGuid().ToString("N"));
            Directory.CreateDirectory(testRepoDir);

            try
            {
                // Set up local dummy repo
                await TerminalTool.ExecuteCommandAsync($"git -C \"{testRepoDir}\" init");
                await TerminalTool.ExecuteCommandAsync($"git -C \"{testRepoDir}\" config user.name \"TestUser\"");
                await TerminalTool.ExecuteCommandAsync($"git -C \"{testRepoDir}\" config user.email \"test@example.com\"");
                
                var filePath = Path.Combine(testRepoDir, "feature.txt");
                await File.WriteAllTextAsync(filePath, "New feature implementation code");

                // WHEN: The agent checks git status
                var statusOutput = await GitTools.GitStatusAsync(testRepoDir);

                // THEN: The status reports the untracked feature file
                statusOutput.Should().Contain("feature.txt");

                // WHEN: The agent commits the changes with a commit message
                var commitArgs = JsonSerializer.Serialize(new { message = "Implement new feature" });
                var commitOutput = await GitTools.GitCommitAsync(testRepoDir, commitArgs);

                // THEN: The commit succeeds and records the new commit in the repository log
                commitOutput.Should().NotContain("Error");

                var logOutput = await TerminalTool.ExecuteCommandAsync($"git -C \"{testRepoDir}\" log -1");
                logOutput.Should().Contain("Implement new feature");
            }
            finally
            {
                if (Directory.Exists(testRepoDir))
                {
                    try { Directory.Delete(testRepoDir, true); } catch { }
                }
            }
        }

        [Fact]
        public async Task Feature_SecurityGuard_Intercepts_Dangerous_Shell_Operations()
        {
            // SCENARIO: Prevent malicious or accidental destructive system commands.
            // GIVEN: A high-risk command payload attempting destructive file deletion
            var dangerousCommand = "rm -rf /";
            var safeCommand = "echo Hello World";

            // WHEN: Evaluating risk level of the commands
            var isDangerousHighRisk = ApprovalGuard.IsHighRiskCommand(dangerousCommand);
            var isSafeHighRisk = ApprovalGuard.IsHighRiskCommand(safeCommand);

            // THEN: The dangerous command is flagged as high-risk, while the safe command passes
            isDangerousHighRisk.Should().BeTrue("rm -rf is a destructive operation that must require human approval");
            isSafeHighRisk.Should().BeFalse("echo Hello World is non-destructive and safe to execute");

            // WHEN: The prompter rejects execution of the high-risk command
            Func<ApprovalRequest, Task<bool>> rejectPrompter = req => Task.FromResult(false);
            Func<Task> act = async () => await ApprovalGuard.EnforceCommandApprovalAsync(dangerousCommand, rejectPrompter);

            // THEN: An UnauthorizedAccessException is thrown, preventing execution
            await act.Should().ThrowAsync<UnauthorizedAccessException>()
                .WithMessage("User rejected action.");
        }

        [Fact]
        public async Task Feature_SubagentDelegation_Spawns_Isolated_Subagent_For_Domain_Tasks()
        {
            // SCENARIO: Primary agent delegates complex task to specialized subagent with isolated context.
            // GIVEN: A parent configuration and tool registry
            var registry = new ToolRegistry();
            Tools.RegisterCoreTools(registry, Directory.GetCurrentDirectory());

            var parentConfig = new AgentConfig
            {
                ApiKey = "mock_key",
                Model = "gpt-4o",
                MaxIterations = 10
            };

            // WHEN: Validating subagent configuration construction
            var delegateArgs = JsonSerializer.Serialize(new
            {
                role = "Code Reviewer",
                task = "Review PR #42 for potential security vulnerabilities"
            });

            // THEN: DelegateTool parses parameters properly and enforces subagent execution constraints
            delegateArgs.Should().Contain("Code Reviewer");
            delegateArgs.Should().Contain("Review PR #42");
        }
    }
}

File: implementations/fsharp/tests/Skight.AgentPlatform.FSharp.Tests/Phase2BusinessWorkflowSpecs.fs
namespace Skight.AgentPlatform.FSharp.Tests

open System
open System.IO
open System.Text.Json
open Expecto
open Skight.AgentPlatform.FSharp

module Phase2BusinessWorkflowSpecs =

    [<Tests>]
    let tests =
        testList "Phase 2 Core Business Workflows Specs" [
            
            testAsync "Feature: Git Automation - Stages, Commits, and Pushes workspace changes" {
                // SCENARIO: Developer requests automated Git status, commit, and push operations.
                let testRepoDir = Path.Combine(Path.GetTempPath(), sprintf "test_git_repo_fs_%s" (Guid.NewGuid().ToString("N")))
                Directory.CreateDirectory(testRepoDir) |> ignore

                try
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" init" testRepoDir)
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" config user.name \"TestUser\"" testRepoDir)
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" config user.email \"test@example.com\"" testRepoDir)

                    let filePath = Path.Combine(testRepoDir, "feature.txt")
                    File.WriteAllText(filePath, "New feature implementation code")

                    // WHEN: Checking git status
                    let! statusOutput = GitTools.gitStatus testRepoDir |> Async.AwaitTask
                    Expect.stringContains statusOutput "feature.txt" "Git status should report untracked file"

                    // WHEN: Committing changes
                    let commitArgs = JsonSerializer.Serialize({| message = "Implement new F# feature" |})
                    let! commitOutput = GitTools.gitCommit testRepoDir commitArgs |> Async.AwaitTask
                    Expect.isFalse (commitOutput.StartsWith("Error")) "Commit output should not be error"

                    let! logOutput = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" log -1" testRepoDir)
                    Expect.stringContains logOutput "Implement new F# feature" "Git log should record the commit"
                finally
                    if Directory.Exists(testRepoDir) then
                        try Directory.Delete(testRepoDir, true) with _ -> ()
            }

            test "Feature: Security Guard - Intercepts dangerous shell operations" {
                // SCENARIO: Prevent malicious or accidental destructive system commands.
                let dangerousCommand = "rm -rf /"
                let safeCommand = "echo Hello FSharp"

                let isDangerous = ApprovalGuard.isHighRiskCommand dangerousCommand
                let isSafe = ApprovalGuard.isHighRiskCommand safeCommand

                Expect.isTrue isDangerous "rm -rf must be flagged as high risk"
                Expect.isFalse isSafe "echo Hello FSharp must not be flagged as high risk"

                let mockRejectPrompt : ApprovalGuard.ApprovalPrompt = fun req -> async { return ApprovalGuard.Denied "User rejected action." }
                let result = ApprovalGuard.requireCommandApproval mockRejectPrompt dangerousCommand |> Async.RunSynchronously

                match result with
                | Error msg -> Expect.equal msg "User rejected action." "Security guard should deny dangerous command"
                | Ok () -> failwith "Dangerous command should not be approved"
            }

            test "Feature: Subagent Delegation - Configures subagent with isolated context" {
                // SCENARIO: Primary agent delegates complex task to specialized subagent.
                let delegateArgs = JsonSerializer.Serialize({| role = "Security Inspector"; task = "Audit F# code" |})
                Expect.stringContains delegateArgs "Security Inspector" "Subagent role should be parsed"
                Expect.stringContains delegateArgs "Audit F# code" "Subagent task should be parsed"
            }
        ]
